sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -588,6 +588,16 @@ object ColumnPruning extends Rule[LogicalPlan] {
         .map(_._2)
       p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices))
 
+    // prune unrequired nested fields
+    case p @ Project(projectList, g: Generate) if SQLConf.get.nestedSchemaPruningEnabled =>
+      NestedColumnAliasing.getAliasSubMap(projectList).map {
+        case (nestedFieldToAlias, attrToAliases) =>
+          val newChild = g.withNewChildren(g.children.map { child =>
+            Project(child.output ++ attrToAliases.values.flatten, child)
```

Comment:
This looks addition instead of a replacing. So, this may optimize `explode`, but we cannot avoid the full scan of underlying parquet files?

Reply:
This is good point. I was planning to deal this as separate issue. As you mentioned and we control it with a pruning on expressions config now, I think it is good to deal with it here. I will update.